### PR TITLE
Add unit tests for explain command

### PR DESCRIPTION
**Pull request by ardaguclu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR adds unit tests for `explain` command to increase unit test coverage of kubectl.

#### Does this PR introduce a user-facing change?
```release-note
None
```

### CODE REVIEW

**Comment by ardaguclu** [comment]
@brianpursley could you please take a look at this?. Thanks.

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	cmd := NewCmdExplain("kubectl", tf, ioStreams)
+	cmd.Run(cmd, []string{"pods"})
+	if !strings.Contains(buf.String(), "KIND:     Pod") {
+		t.Fatalf("expected output should include pod kind")
+	}
+
+	cmd.Flags().Set("recursive", "true")
+	cmd.Run(cmd, []string{"pods"})
+	if !strings.Contains(buf.String(), "KIND:     Pod") ||
+		!strings.Contains(buf.String(), "apiVersion\t<string>") {
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:95]: I don't think this check is unique to `--recursive` as I see `apiVersion\t<string>` in the explain output, even without recursive specified.

```
$ kubectl explain pod | grep -P "apiVersion\t<string>"
   apiVersion	<string>
```

Maybe change this second check to something that will only be in the output when `--recursive` is specified.

For example:  `!strings.Contains(buf.String(), "annotations\t<map[string]string>") {`

WDYT?

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		t.Fatalf("expected output should include pod kind")
+	}
+
+	cmd.Flags().Set("api-version", "v1")
+	cmd.Run(cmd, []string{"pods"})
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:100]: Can you use something that has multiple version defined in the Open API spec and make sure that it shows the correct version?

So for example, I see in `testdata/openapi/swagger.json` that `cronjobs` has both `batch/v1beta1` and `batch/v1` versions defined.

So maybe test it both ways one with `--api-versions=batch/v1beta1` and then with `--api-versions=batch/v1` and look for the second line where is shows like:
```
VERSION:  batch/v1
```

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if !strings.Contains(buf.String(), "KIND:     Pod") ||
+		!strings.Contains(buf.String(), "apiVersion\t<string>") {
+		t.Fatalf("expected output should include pod kind")
+	}
+}
```
Comment (brianpursley) [inline staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:105]: I guess while you are adding unit tests, might as well check the error message when you ask it to explain something it doesn't know about.

So for example:
```
$ kubectl explain foo
the server doesn't have a resource type "foo"
$ echo $?
1
```

And if the resource exists, but the version doesn't it is slightly different:
```
$ kubectl explain pod --api-version=v99
error: couldn't find resource for "/v99, Kind=Pod"
$ echo $?
1
```

**Review by brianpursley — CHANGES_REQUESTED** [review]
Looks like a good start.  I have a few suggestions for some additional tests while you are adding the tests.  Hopefully it won't be too much extra work to put those in.

**Comment by ardaguclu** [comment]
@brianpursley thanks for review. These are all valid and useful points. I made changes according to your suggestions.

**Comment by ardaguclu** [comment]
@brianpursley would you mind looking at one last time to this PR?. Thanks.

**Review by brianpursley — CHANGES_REQUESTED** [review]
Looks good.  Just one comment about a couple of the error messages that I think might refer to the wrong resource type (pod vs cronjob).

File: staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	cmd.Run(cmd, []string{"pods"})
+	if !strings.Contains(buf.String(), "KIND:     Pod") ||
+		!strings.Contains(buf.String(), "annotations\t<map[string]string>") {
+		t.Fatalf("expected output should include pod kind")
+	}
+
+	cmd.Flags().Set("api-version", "batch/v1")
+	cmd.Run(cmd, []string{"cronjobs"})
+	if !strings.Contains(buf.String(), "VERSION:  batch/v1") {
+		t.Fatalf("expected output should include pod kind")
```
[the remark this answers is not in the document]
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/explain/explain_test.go:148]: Or I guess maybe it should say that it expects the version to be batch/v1 which is really what it looks like it is checking

**Comment by ardaguclu** [comment]
unit test failures are not related to this PR

/test pull-kubernetes-unit

**Review by brianpursley — APPROVED** [review]
Waiting for unit tests to pass before official lgtm, but this looks good...

**Comment by ardaguclu** [comment]
> Waiting for unit tests to pass before official lgtm, but this looks good...

unit tests passed :)

**Pull request merged**